plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,38 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+
+			isDNSBufferError := false
+			isDNSOverflowError := false
+			var perr *dns.Error
+
+			if proto == "udp" {
```

Comment:
```suggestion
			if proto == "udp" {
			// For UDP, if the error is an overflow, we probably have an upstream misbehaving in some way.
			// (e.g. sending >512 byte responses without an eDNS0 OPT RR).
			// Instead of returning an error, return an empty response with TC bit set. This will make the
			// client retry over TCP (if that's supported) or at least receive a clean
			// error. The connection is still good so we break before the close.

			isDNSBufferError := false
			isDNSOverflowError := false
			var perr *dns.Error

			
```